Font Face: to generate and print font-face styles for theme.json fonts

hellofromtonya opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Introduce Font Face processing to generate and print `@font-face` styles for fonts in theme.json merged data.

It's minimalistic in what it publicly exposes to avoid BC concerns.

There's no register, deregister, or enqueue of fonts or providers.

The fonts to be processed come directly from theme.json merged data which contains the theme's defined fonts and user "activated" fonts (from the Font Library).

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

See #51769.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

It adds 2 new classes `WP_Font_Face` and `WP_Font_Face_Resolver` which currently live in `/lib/experimental/fonts/` folder. That folder could also house the new Fonts Library.

Jobs:
* The Resolver's job is to handle getting and formatting the font data from theme.json.
* The `WP_Font_Face`'s job is to generate and print the `@font-face` styles.
* The local provider functionality is integrated into `WP_Font_Face`. 

If in the future support for custom providers is needed, the local provider can be removed from the`WP_Font_Face` and reintroduced as a separate class without breaking BC (backwards-compatibility).

When it's time to print, `wp_print_font_faces()` handles a call to the `WP_Font_Face_Resolver::get_fonts_from_theme_json()` and passes the fonts to `WP_Font_Face::print()`.

This also means other sources could pass fonts to `WP_Font_Face::print()`, such as plugins on a classic site.

### Prepared to be merged _before_ Fonts Library

The new files will not be loaded into memory unless the Fonts Library exists. In this way, this PR can be merged into `trunk` before the Fonts Library without breaking websites that are using the Fonts API.

## TODO

- [X] Add `WP_Font_Face_Resolver`.
- [X] Add `WP_Font_Face`.
- [X] Integrate local provider into `WP_Font_Face`.
- [X] Only expose `print()` to hide the internal workings (eliminate future BC concerns).
- [X] Conditional load only when the Fonts Library exists.
- [x] Add PHPUnit tests.

## Testing Instructions

1. Enable the loading of this PR's files by:
    a. Open `lib/load.php`.
    b. Scroll down to where the Font Face files are loaded.
    c. Add `true || ` in the `if`:
```php
if ( true || class_exists( 'WP_Fonts_Library' ) || class_exists( 'WP_Fonts_Library_Controller' ) ) {
```
2. With TT3:
  a. In Site Editor using Dev Tools, check that the `<script id="wp-fonts-local">` and `@font-face` styles are present in `<head>` and the iframed editor.
  b. In a post or page, check that the `<script id="wp-fonts-local">` and `@font-face` styles are present in `<head>` and the iframed editor.
  c. On the front-end, check that the `<script id="wp-fonts-local">` and `@font-face` styles are present.

```html
<style id="wp-fonts-local">
@font-face{font-family:"DM Sans";font-style:normal;font-weight:400;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/dm-sans/DMSans-Regular.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"DM Sans";font-style:italic;font-weight:400;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/dm-sans/DMSans-Regular-Italic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"DM Sans";font-style:normal;font-weight:700;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/dm-sans/DMSans-Bold.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"DM Sans";font-style:italic;font-weight:700;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/dm-sans/DMSans-Bold-Italic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:normal;font-weight:300;font-display:block;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/ibm-plex-mono/IBMPlexMono-Light.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:normal;font-weight:400;font-display:block;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/ibm-plex-mono/IBMPlexMono-Regular.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:italic;font-weight:400;font-display:block;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/ibm-plex-mono/IBMPlexMono-Italic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:normal;font-weight:700;font-display:block;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/ibm-plex-mono/IBMPlexMono-Bold.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:Inter;font-style:normal;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/inter/Inter-VariableFont_slnt,wght.ttf') format('truetype');font-stretch:normal;}
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"Source Serif Pro";font-style:italic;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2') format('woff2');font-stretch:normal;}

</style>
```
2. With TT1:
    * In the `functions.php`, add the following code:
```php
add_action( 'wp_head', 'tt1_test_print_font_faces', 50 );
add_action( 'admin_print_styles', 'tt1_test_print_font_faces', 50 );
function tt1_test_print_font_faces() {
	$fonts = array(
			'Source Serif Pro' => array(
					array(
							'font-family' => 'Source Serif Pro',
							'font-style'   => 'normal',
							'font-weight'  => '200 900',
							'src'          => get_theme_file_uri( 'assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2' ),
					),
					array(
							'font-family' => 'Source Serif Pro',
							'font-style'   => 'italic',
							'font-weight'  => '200 900',
							'src'          => get_theme_file_uri( 'assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2' ),
					),
			),
	);
	wp_print_font_faces( $fonts );
}
```
   * Using Dev Tools, check in the front-end and admin area that the `<script id="wp-fonts-local">` and `@font-face` styles are present.

```html
<style id="wp-fonts-local">
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentyone/assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');}
@font-face{font-family:"Source Serif Pro";font-style:italic;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentyone/assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2') format('woff2');}

</style>
```

hellofromtonya committed:
Add WP_Font_Face_Resolver.

The Resolver handles data interaction between the theme.json
and WP_Font_Face. It gets the fonts from theme.json and converts
into the array structure needed for font-face processing.

hellofromtonya committed:
Add WP_Font_Face.

This new class handles validating font-face properties and
generating and printing font-face styles.

Only the "generate_and_print()" method is public exposed.
All other functionality is hidden as interworkings.

Integrates WP_Fonts_Provider_Local into the class,
rather than as a separate provider class, as Font Face
only supports local font files. By integrating into
the WP_Font_Face, it avoids exposing provider registration
and eliminates future BC concerns. If in the future
support for custom providers is needed, the local
provider functionality can be safely removed from
WP_Font_Face without BC concerns or impacts.

hellofromtonya committed:
Add wp_print_font_faces().

This function works for both block-powered and classic sites.

For block-powered sites, it gets the fonts from theme.json via
the Font Face Resolver.

For classic sites, themes and plugins can programmatically invoke
it and pass their fonts to it.

The instance of WP_Font_Face is variable cached within this
global function instead of providing a separate global function.

Why? The entry point to use Font Face is via wp_print_font_faces().
No other methods are public. Therefore, a global wp_font_face()
function should not be needed as it would not provide benefits.

hellofromtonya committed:
Relocate iframed editor assets font printing.

Removes wp_print_fonts() from _gutenberg_get_iframed_editor_assets().

Adds a filter callback to into each global functions file.

This approach keeps the code centralized with its specific
print function.

Benefits:
* Avoids switching logic in _gutenberg_get_iframed_editor_assets().
* Avoids confusion for backporting to Core.

hellofromtonya committed:
Load files & guard usage of Fonts API.

hellofromtonya committed:
Fix wp_print_font_faces().

1. The action is passing an empty string. Removed the `array` type hint.

2. Fixed the instantiation code. :facepalm:

hellofromtonya committed:
Fixed property typo

hellofromtonya commented:
This PR is ready to be manually tested. I've added test instructions for TT3 (block theme) and TT1 (classic theme). Please test and add test reports.

anton-vlasenko commented:
# Test Report

Env:
* WordPress: 6.3-alpha-55505-src
* PHP: 8.0.27
* Server: Apache/2.4.55 (Unix) PHP/8.0.27
* Database: mysqli (Server: 5.7.38 / Client: mysqlnd 8.0.27)
* Browser: Safari 16.5 (macOS)
* Theme: Twenty Twenty-One 1.8
* MU-Plugins: None activated
* Plugins: 
  * Gutenberg 16.0.0

## Test Instructions

1. Follow the testing instructions listed in the PR's description.

2. Enable TT3, perform the requested code changes and make sure that the `style#wp-fonts-local` element is present and contains all the needed elements, as described in the `Testing Instructions` section (both on frontend and backend).

2. Enable TT1, perform the requested code changes and make sure that the `style#wp-fonts-local` element is present and contains all the needed elements, as described in the `Testing Instructions` section (both on frontend and backend).


## Results

* Yes, the `style#id-fonts-local` is present in `head` and the iframed editor (backend) and on the frontend and backend when using the TT3 theme. ✅
* Yes, the `style#id-fonts-local` is present on the frontend and backend when using the TT1 theme. ✅

anton-vlasenko committed:
Test for Test WP_Font_Face::generate_and_print.

anton-vlasenko committed:
Fix class name.

anton-vlasenko committed:
Fix the test.

anton-vlasenko committed:
Don't load WP_Fonts_TestCase

anton-vlasenko committed:
Don't print the last \n character

anton-vlasenko committed:
Add a comment.

anton-vlasenko committed:
Add more data to the dataprovider.

anton-vlasenko committed:
Rename files.

anton-vlasenko committed:
Fix fatal error.

anton-vlasenko committed:
Fix package name.

anton-vlasenko committed:
Add a draft test for WP_Font_Face_Resolver.

anton-vlasenko committed:
Commit draft test.

anton-vlasenko committed:
Return an empty array if no fonts are defined. This method should always return an array

anton-vlasenko committed:
Improve the test.

aristath committed:
CS fix

aristath commented:
Pushed a codings-standards fix 👍

anton-vlasenko committed:
Fix font names.

anton-vlasenko committed:
Fix package name.

anton-vlasenko committed:
Remove unrelated tests.

anton-vlasenko committed:
Add a test for placeholders.

ndiego commented:
Hi team; where are we on this one? Fonts loaded by theme.json are currently not working in the Editor in WordPress 6.3 Beta 2 when metaboxes exist. I strongly believe this will need to get fixed in 6.3. See https://github.com/WordPress/gutenberg/issues/51209

anton-vlasenko commented:
> Hi team; where are we on this one? Fonts loaded by theme.json are currently not working in the Editor in WordPress 6.3 Beta 2 when metaboxes exist. I strongly believe this will need to get fixed in 6.3. See #51209

@ndiego I think it's almost ready. I just need to fix some unit tests.
I need an hour or so.

UPD: This PR is ready to be reviewed.

anton-vlasenko committed:
Fix property names.
Remove unused property.

anton-vlasenko committed:
Fix DocBlocks.

anton-vlasenko committed:
Fix CS.

anton-vlasenko committed:
Fix test class names.

anton-vlasenko committed:
Tests for wp_print_font_faces().

anton-vlasenko committed:
Improve the test and kick off CI checks.

anton-vlasenko committed:
Fix the failing test.

ndiego commented:
Just tested to see if this [fixes the issue ](https://github.com/WordPress/gutenberg/issues/51209)where fonts registed in theme.json do not load when metaboxes are present in Gutenberg 16.1 or WordPress 6.3 Beta 2 (i.e. the Editor is iframed). Unfortunately, I am still seeing the issue. 

In WordPress 6.2.2, without Gutenberg active, if you enable the Yoast plugin, the fonts in TT3 will load as expected. Yoast includes a metabox and the post Editor in 6.2.2 is not iframed.

Now if you update to 6.3 Beta 2, the fonts are not loaded when a metabox is present (the Editor is no longer iframed).

Reverting back to 6.2.2, if you activate Gutenberg with this PR, unfortunately, the fonts are still not loaded when the metabox is present.

| WordPress 6.2.2 (fonts loaded) | 6.3 Beta 2 or 6.2.2 + this PR (fonts not loaded) |
| --- | --- |
| <img width="1371" alt="image" src="https://github.com/WordPress/gutenberg/assets/4832319/ad174856-514a-4ef2-b6d4-69ff64d4811a"> | <img width="1371" alt="image" src="https://github.com/WordPress/gutenberg/assets/4832319/ebc44ac5-f854-4e47-8214-d78433953ff4"> |

anton-vlasenko commented:
@ndiego this bug is occurring because [this code](https://github.com/WordPress/gutenberg/pull/51770/files#diff-63a02f6d67fa3b923826253ca376613481424de736ac58fafc0d732e34804991R254) attempts to insert font-faces into the `__unstableResolvedAssets` editor setting. However, this setting doesn't work when the editor is not framed.
The solution would be to add font-faces to the `styles` setting instead:
```php
$settings['styles'][] = array(
	'css'            => $styles, // fonts faces are wrapped in the `style#wp-fonts-local` element, and it has to be removed
	'__unstableType' => 'user', // should we use user or theme or something else?
	'isGlobalStyles' => true,
);

return $settings;
```

This solution appears to be functioning correctly, but it requires additional testing.

anton-vlasenko committed:
Quick and dirty fix for non-iframed editor. See https://wordpress.slack.com/archives/C051Z1SKBDZ/p1688392704173649?thread_ts=1688054141.570659&cid=C051Z1SKBDZ

anton-vlasenko commented:
@ndiego 2b61cf206194e1c05afc668d37bf1efb43dc40ec should fix the issue you mentioned above (no fonts in the non-iframed editor mode).

anton-vlasenko committed:
Revert "Quick and dirty fix for non-iframed editor. See https://wordpress.slack.com/archives/C051Z1SKBDZ/p1688392704173649?thread_ts=1688054141.570659&cid=C051Z1SKBDZ"

This reverts commit 2b61cf206194e1c05afc668d37bf1efb43dc40ec.

anton-vlasenko commented:
I reverted my fix ~~because the regression was likely introduced in #46706 (but this needs confirmation).~~

hellofromtonya commented:
>Just tested to see if this https://github.com/WordPress/gutenberg/issues/51209 where fonts registed in theme.json do not load when metaboxes are present in Gutenberg 16.1 or WordPress 6.3 Beta 2 (i.e. the Editor is iframed). Unfortunately, I am still seeing the issue.

Thanks @ndiego and @anton-vlasenko. The issue existed before this PR. IMO it is out-of-scope for fixing in this PR. 

Why is it out-of-scope? This PR seeks to transform the Fonts API into Font Face generator and printer. It's reusing the how-to and code from the Fonts API. It is not seeking to fix existing issues in the Fonts API.

IMO issue https://github.com/WordPress/gutenberg/issues/51209 deserves its own consideration and scope with a separate PR. 

There are many benefits to working in smaller increments, rather than bundling fixes together into larger PRs. These include:

* Capturing everything (including discussions, testing, code reviews, and the commit) for the issue to be solved.
* Provides a historical feed for tracing when something unexpectedly breaks something else, such with a regression.
* The PR should directly address and solve the issue it is attached to.

As this specific PR did not introduce the regression, I asked @anton-vlasenko to revert the commit. I'm seeing where the Fonts API itself directly introduced the regression. I'm reopening the reporting issue. Discussion can continue in the reporting issue.

hellofromtonya commented:
>Hi team; where are we on this one? Fonts loaded by theme.json are currently not working in the Editor in WordPress 6.3 Beta 2 when metaboxes exist. I strongly believe this will need to get fixed in 6.3. See https://github.com/WordPress/gutenberg/issues/51209

The issue in WordPress Core is not related to this PR. Thus https://github.com/WordPress/gutenberg/issues/51209 should be resolved separately.

WordPress 6.3 Beta 2 is still using the stopgap code from WP 6.0.0 (i.e. `_wp_theme_json_webfonts_handler()`). Given that the metabox issue exists in WP 6.3 Beta 2 (which does not have the Fonts API or this new Font Face PR), this suggests the issue was introduced by something else. Thus, IMO the regression work should happen separate from this PR to first identify the merged PR that introduced the regression. Let's shift the conversation back to https://github.com/WordPress/gutenberg/issues/51209.

anton-vlasenko reviewed (commented):
@hellofromtonya When I was writing unit tests for `WP_Font_Face_Resolver::get_fonts_from_theme_json()`, I noticed that it does not return style variations fonts. For example, when using the `fonts-block-theme` test theme, it does not return the `Open Sans` font faces. However, a remotely "similar" method, `WP_Fonts_Resolver::register_fonts_from_theme_json()`, does return the `Open Sans` font faces.
Is this by design?
I'm asking to make sure that this has not been overlooked.

hellofromtonya commented:
>When I was writing unit tests for `WP_Font_Face_Resolver::get_fonts_from_theme_json()`, I noticed that it does not return style variations fonts. For example, when using the fonts-block-theme test theme, it does not return the Open Sans font faces. However, a remotely "similar" method, `WP_Fonts_Resolver::register_fonts_from_theme_json()`, does return the Open Sans font faces.
>Is this by design?

@anton-vlasenko good question. Yes, it's by design. 

Why? 

The Fonts API pulled all theme defined fonts (i.e. in the main `theme.json` and style variation files) to get all of them registered. It wasn't necessarily smart about identifying which fonts should be used as those decisions were made higher up at the user level.

The Font Face though is designed to only use the Theme JSON merged data layers. All fonts that need to be generated and printed are within that merged data.

What if a user selects a style variation? Saving it should bring any fonts defined in the style variation into the "user" data layer, which is part of the merged data the Font Face uses. The UI uses the merged data layer for populating the font pickers. Thus, this is the same data the Font Face should use to keep parity.

If a user selects a style variation, are fonts in that style variation added into the merged data layer? Yes, they should be. I would be good to verify this, though that's out-of-scope for this PR. If there's a bug, the fix likely resides in the Theme JSON Resolver, rather than in the Font Face (i.e. to ensure all consumers are using the same data - font pickers, font library, font face, etc.).

hellofromtonya commented:
@anton-vlasenko if you're happy with the PR, please go ahead and approve it.

I'll do a code review of the PHPUnit tests. That way, you've code reviewed the source code changes and I've code reviewed the tests. Teamwork!

anton-vlasenko commented:
> I'll do a code review of the PHPUnit tests. That way, you've code reviewed the source code changes and I've code reviewed the tests. Teamwork!

Awesome, @hellofromtonya. Thanks!

> if you're happy with the PR, please go ahead and approve it.

I'd like to rebase this PR onto trunk to see if it works with the latests fixes (i.e. https://github.com/WordPress/gutenberg/issues/51209) and test it again (and probably add another test report).
If everything is fine, I think I will approve it tomorrow.
Would that be fine with you, @hellofromtonya?

hellofromtonya commented:
Rebased on top of `trunk` to pull in the latest changes ✅

hellofromtonya committed:
Test reorganization.

1. Adds a test case to the fonts directory.

2. Adds a dataset trait for sharing datasets amongst different test classes.
This change the maintance burden of redundant datasets by centralizing in a shared trait.

3.Removes the style variation tests as these are no longer relevant to the Font Face.

4. A wee bit of renaming and reformatting.

hellofromtonya committed:
Add wp_print_fonts() tests.

1. Adds test for when no fonts are passed into the function, i.e. meaning it will pull from the resolver.

2. Improves resolver test to validate the full returned array structure.

3. Moves the expected fonts to the dataset trait to share it between these 2 tests.

4. A wee bit of girlscouting.

hellofromtonya committed:
Bail out if no fonts to process.

1. Adds guard clause to WP_Font_Face::generate_and_print().
2. Adds unhappy test for it.
3. Adds similar unhappy test for wp_print_font_faces().

hellofromtonya commented:
@anton-vlasenko I reviewed the tests. You did a great job! Thank you :) To help, I made the changes from the review in the 3 added commits with commit messages explaining the changes. What do you think?

anton-vlasenko commented on lib/experimental/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,418 @@
+			&&
+			function_exists( 'current_theme_supports' ) && ! current_theme_supports( 'html5', 'style' )
+		) {
+			$this->style_tag_attrs = array( 'type' => 'text/css' );
```
The `type` attribute is deprecated, as stated in https://developer.mozilla.org/en-US/docs/Web/HTML/Element/style#deprecated_attributes.
However, I don't think this could create any issues (and the `type` attribute can be removed later if absolutely necessary).

hellofromtonya replied:
Core still supports pre HTML 5 syntax. This particular code comes from WP Styles. Once Core drops < HTML 5, then this can also be removed.

anton-vlasenko reviewed (approved):
I have tested this PR again, and [my test report](https://github.com/WordPress/gutenberg/pull/51770#issuecomment-1604921628) is still valid considering the recent fixes.

Additionally, I tested different combinations, such as meta boxes enabled/disabled, block/classical theme, and I couldn't find any issues (to the best of my knowledge).

The code also looks good to me.
Therefore, I approve this PR.

hellofromtonya commented:
Thank you everyone for your contributions :) I've merged this PR into `trunk`.

>The new files will not be loaded into memory unless the Fonts Library exists. In this way, this PR can be merged into trunk before the Fonts Library without breaking websites that are using the Fonts API.

Remember, Font Face only loads into memory (replacing the Fonts API) _when_ the Fonts Library is also loaded into memory. That's by design to keep sites running (using Fonts API) while providing the new typography path forward for testing with the Fonts Library.

hellofromtonya commented:
Removed the `Needs PHP backport` label as Font Face was merged into Core via changeset https://core.trac.wordpress.org/changeset/56500 ✅

ramonjd commented on lib/experimental/fonts/class-wp-font-face-resolver.php:
```diff
@@ -0,0 +1,158 @@
+
+			// Converts the "file:./" src placeholder into a theme font file URI.
+			if ( ! empty( $font_face['src'] ) ) {
+				$font_face['src'] = static::to_theme_file_uri( (array) $font_face['src'] );
```
I know this is in Core already, but it relates to a feature being built for 6.6: [relative paths to background images in theme.json styles](https://github.com/WordPress/gutenberg/pull/61271)

For consistency, and after a nudge from @creativecoder, I decided to follow the `file:./` dotslash path convention to represent paths to these images.

@noisysocks then pointed out that dotslash indicates the current directory, and might imply other dot-style paths, e.g., `../`

So, if web fonts can be used in theme variations, e.g., `my-theme/styles/variation-with-fonts.json`, the `./` would indicate a path relative to the styles directory.

It might also imply that paths with `../../some/path.file` could also be resolved. But I'm not sure if `get_theme_file_uri()` supports such things as it looks for files relative to the active or parent theme root. Maybe I'm missing something.

https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.4/fonts/font-face/class-wp-font-face-resolver.php#L147C27-L147C44

I was wondering if we should enforce a convention. Either:

1. Enforce that all paths must be relative to the theme root, or
2. Allow relative to current directory, and then in resolve them in the background with PHP trickery with [realpath](https://www.php.net/manual/en/function.realpath.php) or something.

What do folks think?

creativecoder replied:
I see what you mean... the syntax and behavior seem mismatched: `file:./` seems to imply a reference to the current directory, but really it's a placeholder for the theme root looking at how it's coded.

Updating `file:./` to apply to the current directory (e.g. `theme-root/styles`) for theme variations and/or adding resolution of `../` paths would make the syntax clearer. But I worry about backward compatibility (I imagine that some theme variations are already using `file:./`?). And some care is needed to prevent directory traversal outside the theme directory.

So I favor keeping things simple, and continuing to use `file:./` simply as a placeholder to the theme directory root. But it's not a strong opinion.

(Note that either way, we should add to and update the [theme handbook docs](https://developer.wordpress.org/themes/global-settings-and-styles/settings/typography/#registering-web-fonts-font-faces) to make it clearer).

ramonjd replied:
Thanks for the quick reply @creativecoder 

> So I favor keeping things simple, and continuing to use file:./ simply as a placeholder to the theme directory root. 

👍🏻 

I went with this approach for now given that folks might be using to it already. 

> Updating file:./ to apply to the current directory (e.g. theme-root/styles) for theme variations and/or adding resolution of ../ paths would make the syntax clearer. But I worry about backward compatibility (I imagine that some theme variations are already using file:./?).

True. Backwards compatibility is now a requirement regardless. And I considered this in https://github.com/WordPress/gutenberg/pull/61271 and contemplated supporting `file:./` (with the `./` for folks who are now used to it) and also a `file:` prefix. 

But I agree that it's simpler to go with what's there until folks come up with a viable alternative. Even if it's _"Enforce that all paths must be relative to the theme root"_ - that way, we could update the docs now and make the `./` optional later without adding any complex `../` support etc. 🤷🏻
